### PR TITLE
feat: add mempool min fee

### PR DESCRIPTION
**Pull request by SWvheerden**
Description
---
adds in a config option to specify the min fee accepted by the mempool

Motivation and Context
---
It is possible to flood the mempool with multiple cheap low fee transactions and waste resources validating them. 

How Has This Been Tested?
---
Unit test

Audit Finding Number
---
TARI-004

### CODE REVIEW

File: applications/minotari_node/src/grpc/base_node_grpc_server.rs [hunk 1]
```diff
@@ -716,6 +716,7 @@ impl tari_rpc::base_node_server::BaseNode for BaseNodeGrpcServer {
             TxStorageResponse::NotStored |
             TxStorageResponse::NotStoredOrphan |
             TxStorageResponse::NotStoredConsensus |
+            TxStorageResponse::NotStoredFeeToLow |
```
Comment (CjS77) [inline applications/minotari_node/src/grpc/base_node_grpc_server.rs]:
```suggestion
            TxStorageResponse::NotStoredFeeTooLow |
```

File: applications/minotari_node/src/grpc/base_node_grpc_server.rs [hunk 2]
```diff
@@ -787,6 +788,7 @@ impl tari_rpc::base_node_server::BaseNode for BaseNodeGrpcServer {
             TxStorageResponse::NotStored |
             TxStorageResponse::NotStoredConsensus |
             TxStorageResponse::NotStoredOrphan |
+            TxStorageResponse::NotStoredFeeToLow |
```
Comment (CjS77) [inline applications/minotari_node/src/grpc/base_node_grpc_server.rs]:
```suggestion
            TxStorageResponse::NotStoredFeeTooLow |
```

File: base_layer/core/src/base_node/proto/wallet_rpc.proto [hunk 3]
```diff
@@ -17,6 +17,7 @@ enum TxSubmissionRejectionReason {
   TxSubmissionRejectionReasonOrphan = 3;
   TxSubmissionRejectionReasonTimeLocked = 4;
   TxSubmissionRejectionReasonValidationFailed = 5;
+  TxSubmissionRejectionReasonFeeToLow = 6;
```
Comment (CjS77) [inline base_layer/core/src/base_node/proto/wallet_rpc.proto]:
```suggestion
  TxSubmissionRejectionReasonFeeTooLow = 6;
```

File: base_layer/core/src/base_node/proto/wallet_rpc.rs [hunk 4]
```diff
@@ -45,6 +45,7 @@ pub enum TxSubmissionRejectionReason {
     Orphan,
     TimeLocked,
     ValidationFailed,
+    FeeToLow,
```
Comment (CjS77) [inline base_layer/core/src/base_node/proto/wallet_rpc.rs]:
```suggestion
    FeeTooLow,
```

File: base_layer/core/src/base_node/proto/wallet_rpc.rs [hunk 5]
```diff
@@ -57,6 +58,7 @@ impl Display for TxSubmissionRejectionReason {
             Orphan => "Orphan",
             TimeLocked => "Time Locked",
             ValidationFailed => "Validation Failed",
+            FeeToLow => "Fee to low",
```
Comment (CjS77) [inline base_layer/core/src/base_node/proto/wallet_rpc.rs]:
```suggestion
            FeeTooLow => "Fee too low",
```

File: base_layer/core/src/base_node/proto/wallet_rpc.rs [hunk 6]
```diff
@@ -76,6 +78,7 @@ impl TryFrom<proto::TxSubmissionRejectionReason> for TxSubmissionRejectionReason
             Orphan => TxSubmissionRejectionReason::Orphan,
             TimeLocked => TxSubmissionRejectionReason::TimeLocked,
             ValidationFailed => TxSubmissionRejectionReason::ValidationFailed,
+            FeeToLow => TxSubmissionRejectionReason::FeeToLow,
```
Comment (CjS77) [inline base_layer/core/src/base_node/proto/wallet_rpc.rs]:
```suggestion
            FeeTooLow => TxSubmissionRejectionReason::FeeTooLow,
```

File: base_layer/core/src/base_node/proto/wallet_rpc.rs [hunk 7]
```diff
@@ -91,6 +94,7 @@ impl From<TxSubmissionRejectionReason> for proto::TxSubmissionRejectionReason {
             Orphan => proto::TxSubmissionRejectionReason::Orphan,
             TimeLocked => proto::TxSubmissionRejectionReason::TimeLocked,
             ValidationFailed => proto::TxSubmissionRejectionReason::ValidationFailed,
+            FeeToLow => proto::TxSubmissionRejectionReason::FeeToLow,
```
Comment (CjS77) [inline base_layer/core/src/base_node/proto/wallet_rpc.rs]:
```suggestion
            FeeTooLow => proto::TxSubmissionRejectionReason::FeeTooLow,
```

File: base_layer/core/src/base_node/rpc/service.rs [hunk 8]
```diff
@@ -150,11 +158,12 @@ impl<B: BlockchainBackend + 'static> BaseNodeWalletRpcService<B> {
             TxStorageResponse::NotStoredAlreadySpent |
             TxStorageResponse::NotStoredConsensus |
             TxStorageResponse::NotStored |
+            TxStorageResponse::NotStoredFeeToLow |
```
Comment (CjS77) [inline base_layer/core/src/base_node/rpc/service.rs]:
```suggestion
            TxStorageResponse::NotStoredFeeTooLow |
```

File: base_layer/core/src/base_node/rpc/service.rs [hunk 9]
```diff
@@ -198,6 +207,11 @@ impl<B: BlockchainBackend + 'static> BaseNodeWalletService for BaseNodeWalletRpc
                 rejection_reason: TxSubmissionRejectionReason::Orphan.into(),
                 is_synced,
             },
+            TxStorageResponse::NotStoredFeeToLow => TxSubmissionResponse {
+                accepted: false,
+                rejection_reason: TxSubmissionRejectionReason::FeeToLow.into(),
+                is_synced,
+            },
```
Comment (CjS77) [inline base_layer/core/src/base_node/rpc/service.rs]:
```suggestion
            TxStorageResponse::NotStoredFeeTooLow => TxSubmissionResponse {
                accepted: false,
                rejection_reason: TxSubmissionRejectionReason::FeeTooLow.into(),
                is_synced,
            },
```

File: base_layer/core/src/mempool/mempool_storage.rs [hunk 10]
```diff
@@ -88,6 +88,10 @@ impl MempoolStorage {
                 );
                 let timer = Instant::now();
                 let weight = self.get_transaction_weighting();
+                if tx.body.get_total_fee().as_u64() < self.unconfirmed_pool.config.min_fee {
+                    warn!(target: LOG_TARGET, "Tx: ({}) fee too low, rejecting",tx_id);
+                    return Ok(TxStorageResponse::NotStoredFeeToLow);
```
Comment (CjS77) [inline base_layer/core/src/mempool/mempool_storage.rs]:
```suggestion
                    return Ok(TxStorageResponse::NotStoredFeeTooLow);
```

File: base_layer/core/src/mempool/mempool_storage.rs [hunk 11]
```diff
@@ -100,6 +104,10 @@ impl MempoolStorage {
             Err(ValidationError::UnknownInputs(dependent_outputs)) => {
                 if self.unconfirmed_pool.contains_all_outputs(&dependent_outputs) {
                     let weight = self.get_transaction_weighting();
+                    if tx.body.get_total_fee().as_u64() < self.unconfirmed_pool.config.min_fee {
+                        warn!(target: LOG_TARGET, "Tx: ({}) fee too low, rejecting",tx_id);
+                        return Ok(TxStorageResponse::NotStoredFeeToLow);
```
Comment (CjS77) [inline base_layer/core/src/mempool/mempool_storage.rs]:
```suggestion
                        return Ok(TxStorageResponse::NotStoredFeeTooLow);
```

File: base_layer/core/src/mempool/mempool_storage.rs [hunk 12]
```diff
@@ -88,6 +88,10 @@ impl MempoolStorage {
                 );
                 let timer = Instant::now();
                 let weight = self.get_transaction_weighting();
+                if tx.body.get_total_fee().as_u64() < self.unconfirmed_pool.config.min_fee {
+                    warn!(target: LOG_TARGET, "Tx: ({}) fee too low, rejecting",tx_id);
```
Comment (CjS77) [inline base_layer/core/src/mempool/mempool_storage.rs]: Warn is too high. It's not a threat if the fees are too low, and you don't want the warning channel to be spammed if someone is flooding with zero-fee txs.
Info, or even debug should be fine.

File: base_layer/core/src/mempool/mempool_storage.rs [hunk 13]
```diff
@@ -100,6 +104,10 @@ impl MempoolStorage {
             Err(ValidationError::UnknownInputs(dependent_outputs)) => {
                 if self.unconfirmed_pool.contains_all_outputs(&dependent_outputs) {
                     let weight = self.get_transaction_weighting();
+                    if tx.body.get_total_fee().as_u64() < self.unconfirmed_pool.config.min_fee {
+                        warn!(target: LOG_TARGET, "Tx: ({}) fee too low, rejecting",tx_id);
```
Comment (CjS77) [inline base_layer/core/src/mempool/mempool_storage.rs]: ditto. Info or debug is a better choice.

File: base_layer/core/src/mempool/mod.rs [hunk 14]
```diff
@@ -114,6 +114,7 @@ pub enum TxStorageResponse {
     NotStoredConsensus,
     NotStored,
     NotStoredAlreadyMined,
+    NotStoredFeeToLow,
```
Comment (CjS77) [inline base_layer/core/src/mempool/mod.rs]:
```suggestion
    NotStoredFeeTooLow,
```

File: base_layer/core/src/mempool/mod.rs [hunk 15]
```diff
@@ -133,6 +134,7 @@ impl Display for TxStorageResponse {
             TxStorageResponse::NotStoredConsensus => "Not stored due to consensus rule",
             TxStorageResponse::NotStored => "Not stored",
             TxStorageResponse::NotStoredAlreadyMined => "Not stored tx already mined",
+            TxStorageResponse::NotStoredFeeToLow => "Not stored tx fee is below the minimum accepted by this mempool",
```
Comment (CjS77) [inline base_layer/core/src/mempool/mod.rs]:
```suggestion
            TxStorageResponse::NotStoredFeeTooLow => "Not stored tx fee is below the minimum accepted by this mempool",
```

File: base_layer/core/src/mempool/proto/tx_storage_response.rs [hunk 16]
```diff
@@ -51,6 +51,7 @@ impl From<TxStorageResponse> for proto::TxStorageResponse {
             NotStoredAlreadySpent => proto::TxStorageResponse::NotStored,
             NotStoredConsensus => proto::TxStorageResponse::NotStored,
             NotStoredAlreadyMined => proto::TxStorageResponse::NotStored,
+            NotStoredFeeToLow => proto::TxStorageResponse::NotStored,
```
Comment (CjS77) [inline base_layer/core/src/mempool/proto/tx_storage_response.rs]:
```suggestion
            NotStoredFeeTooLow => proto::TxStorageResponse::NotStored,
```

**Review by CjS77 — COMMENTED** [review]
Some nits.

to != too.

Log level suggestion.

File: common/config/presets/c_base_node.toml [hunk 17]
```diff
@@ -107,6 +107,8 @@ track_reorgs = true
 # The maximum number of transactions that can be skipped when compiling a set of highest priority transactions,
 # skipping over large transactions are performed in an attempt to fit more transactions into the remaining space.
 #unconfirmed_pool.weight_tx_skip_count = 20
+# The minimum fee accepted by the mempool
+#unconfirmed_pool.min_fee: = 0,
```
Comment (stringhandler) [inline common/config/presets/c_base_node.toml]:
```suggestion
#unconfirmed_pool.min_fee = 0,
```

**Review by stringhandler — DISMISSED** [review]

**Pull request merged**